resources/fetchersManager/data.go

```diff
@@ -165,9 +166,16 @@ func (d *Data) fetchProtected(ctx context.Context, k string, metadata fetching.C
 // Stop cleans up Data resources gracefully.
 func (d *Data) stopData(ctx context.Context, grace time.Duration) {
 	d.stopNotice <- grace
-	ctx, cancel := context.WithTimeout(ctx, grace)
-	defer cancel()
-	<-ctx.Done()
-	d.fetchers.Stop()
-	close(d.stop)
+
+	defer func() {
+		d.fetchers.Stop()
+		close(d.stop)
+	}()
+
+	select {
+	case <-time.After(grace):
+		return
+	case <-ctx.Done():
+		return
+	}
```

Comment:
wait for grace period or context to be cancelled, whichever comes first. context will get canceled if a new fetch iteration started after a grace notice was received.